scripts/west_commands/build_helpers.py

```diff
@@ -13,35 +13,70 @@
 import zcmake
 import os
 from west import log
+from west.configuration import config
+from west.util import escapes_directory
 
 DEFAULT_BUILD_DIR = 'build'
 '''Name of the default Zephyr build directory.'''
 
 DEFAULT_CMAKE_GENERATOR = 'Ninja'
 '''Name of the default CMake generator.'''
 
-BUILD_DIR_DESCRIPTION = '''\
-Build directory. If missing and run in a Zephyr build directory, it is
-used; otherwise, it's "{}".'''.format(
-    DEFAULT_BUILD_DIR)
+FIND_BUILD_DIR_DESCRIPTION = '''\
+If not given, the default build directory ({}/ unless the
+build.dir-fmt configuration variable is set) and the current directory are
+checked, in that order. If one is a Zephyr build directory, it is used.
+'''.format(DEFAULT_BUILD_DIR)
 
+def _resolve_build_dir(fmt, cwd, **kwargs):
+    # Remove any None values, we do not want 'None' as a string
+    kwargs = {k:v for k,v in kwargs.items() if v is not None}
+    # Check if source_dir is below cwd first
+    source_dir = kwargs.get('source_dir')
+    if source_dir:
+        if escapes_directory(cwd, source_dir):
+            kwargs['source_dir'] = os.path.relpath(source_dir, cwd)
+        else:
+            # no meaningful relative path possible
+            kwargs['source_dir'] = ''
+
+    return fmt.format(**kwargs)
 
-def find_build_dir(dir):
+def find_build_dir(dir, **kwargs):
     '''Heuristic for finding a build directory.
 
+    The default build directory is computed by reading the build.dir-fmt
+    configuration option, defaulting to DEFAULT_BUILD_DIR if not set. It might
+    be None if the build.dir-fmt configuration option is set but cannot be
+    resolved.
     If the given argument is truthy, it is returned. Otherwise, if
-    the current working directory is a build directory, it is
-    returned. Otherwise, DEFAULT_BUILD_DIR is returned.'''
+    the default build folder is a build directory, it is returned.
+    Next, if the current working directory is a build directory, it is
+    returned. Finally, the default build directory is returned (may be None).
+    '''
+
     if dir:
         build_dir = dir
     else:
         cwd = os.getcwd()
-        if is_zephyr_build(cwd):
+        default = config.get('build', 'dir-fmt', fallback=DEFAULT_BUILD_DIR)
+        try:
+            default = _resolve_build_dir(default, cwd, **kwargs)
+            log.dbg('config dir-fmt: {}'.format(default),
+                    level=log.VERBOSE_EXTREME)
+        except KeyError as k:
```

Comment:
`as k` with `k` unused is a flake8 warning; to silence it, use `except KeyError:` instead